File: claude_long/python/feature_manager.py
"""
Feature Manager - 功能列表管理器 (Python 实现)

管理 feature_list.json 文件，追踪项目功能的实现状态。
"""

import json
import os
from datetime import datetime
from typing import List, Dict, Optional


class FeatureManager:
    """管理功能列表的类"""

    def __init__(self, feature_file: str = "feature_list.json"):
        self.feature_file = feature_file

    def initialize(self, features: List[Dict] = None) -> None:
        """初始化功能列表文件"""
        if os.path.exists(self.feature_file):
            print(f"Warning: {self.feature_file} already exists")
            return

        if features is None:
            features = [{
                "id": "F001",
                "category": "setup",
                "priority": 1,
                "description": "Project initialization",
                "steps": ["Initialize structure", "Setup environment"],
                "passes": False,
                "notes": "",
                "created_at": datetime.now().isoformat()
            }]

        self.save(features)
        print(f"Initialized feature list: {self.feature_file}")

    def load(self) -> List[Dict]:
        """加载功能列表"""
        if not os.path.exists(self.feature_file):
            return []
        with open(self.feature_file, 'r', encoding='utf-8') as f:
            return json.load(f)

    def save(self, features: List[Dict]) -> None:
        """保存功能列表"""
        with open(self.feature_file, 'w', encoding='utf-8') as f:
            json.dump(features, f, indent=2, ensure_ascii=False)

    def get_next_pending_feature(self) -> Optional[Dict]:
        """获取下一个待完成的功能"""
        features = self.load()
        pending = [f for f in features if not f.get('passes', False)]
        if not pending:
            return None
        pending.sort(key=lambda x: x.get('priority', 999))
        return pending[0]

    def mark_as_complete(self, feature_id: str, notes: str = "") -> bool:
        """标记功能为完成"""
        features = self.load()
        for feature in features:
            if feature['id'] == feature_id:
                feature['passes'] = True
                feature['completed_at'] = datetime.now().isoformat()
                if notes:
                    feature['notes'] = notes
                self.save(features)
                print(f"Marked feature {feature_id} as complete")
                return True
        print(f"Feature {feature_id} not found")
        return False

    def get_summary(self) -> Dict:
        """获取功能列表摘要"""
        features = self.load()
        completed = [f for f in features if f.get('passes', False)]
        return {
            'total': len(features),
            'completed': len(completed),
            'pending': len(features) - len(completed),
            'completion_rate': len(completed) / len(features) if features else 0
        }


if __name__ == "__main__":
    import sys
    if len(sys.argv) < 2:
        print("Usage: python feature_manager.py <command>")
        print("Commands: init, list, next, complete, summary")
        sys.exit(1)

    fm = FeatureManager()
    command = sys.argv[1]

    if command == "init":
        fm.initialize()
    elif command == "list":
        for f in fm.load():
            status = "✓" if f.get('passes') else "○"
            print(f"{status} [{f['id']}] {f['description']}")
    elif command == "next":
        feature = fm.get_next_pending_feature()
        if feature:
            print(f"Next: [{feature['id']}] {feature['description']}")
        else:
            print("All features completed!")
    elif command == "complete":
        if len(sys.argv) < 3:
            print("Usage: complete <feature_id>")
        else:
            fm.mark_as_complete(sys.argv[2])
    elif command == "summary":
        summary = fm.get_summary()
        print(f"Total: {summary['total']}, Completed: {summary['completed']}, Pending: {summary['pending']}")

File: claude_long/go/feature_manager.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"time"
)

// Feature 表示一个功能项
type Feature struct {
	ID          string   `json:"id"`
	Category    string   `json:"category"`
	Priority    int      `json:"priority"`
	Description string   `json:"description"`
	Steps       []string `json:"steps"`
	Passes      bool     `json:"passes"`
	Notes       string   `json:"notes"`
	CreatedAt   string   `json:"created_at"`
	CompletedAt string   `json:"completed_at,omitempty"`
}

// FeatureManager 管理功能列表
type FeatureManager struct {
	FeatureFile string
}

// NewFeatureManager 创建新的功能管理器
func NewFeatureManager(featureFile string) *FeatureManager {
	if featureFile == "" {
		featureFile = "feature_list.json"
	}
	return &FeatureManager{FeatureFile: featureFile}
}

// Initialize 初始化功能列表文件
func (fm *FeatureManager) Initialize(features []Feature) error {
	if _, err := os.Stat(fm.FeatureFile); err == nil {
		fmt.Printf("Warning: %s already exists\n", fm.FeatureFile)
		return nil
	}

	if features == nil {
		features = []Feature{
			{
				ID:          "F001",
				Category:    "setup",
				Priority:    1,
				Description: "Project initialization",
				Steps:       []string{"Initialize structure", "Setup environment"},
				Passes:      false,
				Notes:       "",
				CreatedAt:   time.Now().Format(time.RFC3339),
			},
		}
	}

	return fm.Save(features)
}

// Load 加载功能列表
func (fm *FeatureManager) Load() ([]Feature, error) {
	data, err := os.ReadFile(fm.FeatureFile)
	if err != nil {
		if os.IsNotExist(err) {
			return []Feature{}, nil
		}
		return nil, err
	}

	var features []Feature
	err = json.Unmarshal(data, &features)
	return features, err
}

// Save 保存功能列表
func (fm *FeatureManager) Save(features []Feature) error {
	data, err := json.MarshalIndent(features, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(fm.FeatureFile, data, 0644)
}

// GetNextPendingFeature 获取下一个待完成的功能
func (fm *FeatureManager) GetNextPendingFeature() (*Feature, error) {
	features, err := fm.Load()
	if err != nil {
		return nil, err
	}

	// 过滤未完成的功能
	var pending []Feature
	for _, f := range features {
		if !f.Passes {
			pending = append(pending, f)
		}
	}

	if len(pending) == 0 {
		return nil, nil
	}

	// 按优先级排序
	sort.Slice(pending, func(i, j int) bool {
		return pending[i].Priority < pending[j].Priority
	})

	return &pending[0], nil
}

// MarkAsComplete 标记功能为完成
func (fm *FeatureManager) MarkAsComplete(featureID, notes string) error {
	features, err := fm.Load()
	if err != nil {
		return err
	}

	found := false
	for i := range features {
		if features[i].ID == featureID {
			features[i].Passes = true
			features[i].CompletedAt = time.Now().Format(time.RFC3339)
			if notes != "" {
				features[i].Notes = notes
			}
			found = true
			break
		}
	}

	if !found {
		return fmt.Errorf("feature %s not found", featureID)
	}

	err = fm.Save(features)
	if err != nil {
		return err
	}

	fmt.Printf("Marked feature %s as complete\n", featureID)
	return nil
}

// GetSummary 获取功能列表摘要
func (fm *FeatureManager) GetSummary() (map[string]interface{}, error) {
	features, err := fm.Load()
	if err != nil {
		return nil, err
	}

	completed := 0
	for _, f := range features {
		if f.Passes {
			completed++
		}
	}

	completionRate := 0.0
	if len(features) > 0 {
		completionRate = float64(completed) / float64(len(features))
	}

	return map[string]interface{}{
		"total":           len(features),
		"completed":       completed,
		"pending":         len(features) - completed,
		"completion_rate": completionRate,
	}, nil
}
